core/src/main/scala/neotypes/Session.scala

```diff
@@ -44,7 +50,30 @@ object Session {
         case (tx, Some(_)) => tx.rollback
       }
 
-    override final def close: F[Unit] =
+    override final def readOnlyTransact[T](txf: Transaction[F] => F[T]): F[T] =
+      readOnlyTransact(NeoTransactionConfig.empty())(txf)
+
+    override final def readOnlyTransact[T](config: NeoTransactionConfig)(txf: Transaction[F] => F[T]): F[T] =
+      lock.acquire.flatMap { _ =>
+        F.async[F[T]] { cb =>
+          val work: AsyncTransactionWork[CompletionStage[Unit]] = (tx: NeoAsyncTransaction) => {
+            val cf = new CompletableFuture[Unit]()
+
+            cb(Right(
+              F.delay(Transaction(F, tx)(lock)).guarantee(txf) {
+                case (tx, None)    => tx.commit.flatMap(_ => F.delay(internal.utils.void(cf.complete(()))))
+                case (tx, Some(_)) => tx.rollback.flatMap(_ => F.delay(internal.utils.void(cf.complete(()))))
+              }
```

Comment:
#164 will introduce a helper function called `txFinalizer` which we should use here like `.guarantee(txf)(txFinalizer.andThen(_.flatMap(_ => F.delay(internal.utils.void(cf.complete(()))))))`